sql-parser/src/main/java/io/crate/sql/tree/FrameBound.java

```diff
@@ -23,15 +23,125 @@
 package io.crate.sql.tree;
 
 import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.List;
+
+import static io.crate.common.collections.Lists2.arePeers;
+import static io.crate.common.collections.Lists2.findFirstNonPeer;
 
 public class FrameBound extends Node {
 
     public enum Type {
-        UNBOUNDED_PRECEDING,
-        PRECEDING,
-        CURRENT_ROW,
-        FOLLOWING,
-        UNBOUNDED_FOLLOWING
+        UNBOUNDED_PRECEDING {
+            @Override
+            public <T> int getStart(int pStart,
+                                    int pEnd,
+                                    int currentFrameStartIdx,
+                                    int currentRowIdx,
+                                    @Nullable Comparator<T> cmp,
+                                    List<T> rows) {
+                return pStart;
+            }
+
+            @Override
+            public <T> int getEnd(int pStart, int pEnd, int currentRowIdx, @Nullable Comparator<T> cmp, List<T> rows) {
+                throw new IllegalStateException("UNBOUNDED PRECEDING cannot be the start of a frame");
+            }
+        },
+        PRECEDING {
+            @Override
+            public <T> int getStart(int pStart,
+                                    int pEnd,
+                                    int currentFrameStartIdx,
+                                    int currentRowIdx,
+                                    @Nullable Comparator<T> cmp,
+                                    List<T> rows) {
+                throw new UnsupportedOperationException("Custom PRECEDING frames are not supported");
+            }
+
+            @Override
+            public <T> int getEnd(int pStart, int pEnd, int currentRowIdx, @Nullable Comparator<T> cmp, List<T> rows) {
+                throw new UnsupportedOperationException("Custom PRECEDING frames are not supported");
+            }
+        },
+        /*
+         * In RANGE mode:
+         *    - Frame starts with the current row's first peer (Row that is equal based on the ORDER BY clause)
+         *    - Frame ends with the current row's first peer
+         * In ROWS mode:
+         *    - The current row
+         */
+        CURRENT_ROW {
+            @Override
+            public <T> int getStart(int pStart,
+                                    int pEnd,
+                                    int currentFrameStartIdx,
```

Comment:
Could does arg now also be removed and instead the first peer calculation be done based on `currentRowIdx` and `rows` ?

Reply:
yes, good shout! https://github.com/crate/crate/pull/8759/commits/84d9498b29fb681aa5db9f186c1cd7a7a09fb016